### PR TITLE
Ported to Fabirc 1.17

### PR DESCRIPTION
**Pull request by lucsoft**
Only Document.render is still missing 

Then its merge ready and can directly be published for the public

### CODE REVIEW

File: src/main/java/li/cil/tis3d/api/util/RenderLayerAccess.java [hunk 1]
```diff
@@ -4,42 +4,49 @@
...
+                .transparency(NO_TRANSPARENCY)
+                .depthTest(RenderPhase.LEQUAL_DEPTH_TEST)
+                .cull(RenderPhase.DISABLE_CULLING)
+                .lightmap(ENABLE_LIGHTMAP)
+                .overlay(ENABLE_OVERLAY_COLOR)
+                .build(true);
+
+            return of(CUTOUT_NO_DIFFUSE_LIGHT_LAYER_NAME,
+                VertexFormats.POSITION_COLOR_TEXTURE_OVERLAY_LIGHT_NORMAL,
+                VertexFormat.DrawMode.QUADS, 256, false, false, parameters);
```
Comment (Sturmlilie) [inline src/main/java/li/cil/tis3d/api/util/RenderLayerAccess.java:47]: What happened here? Is the custom RenderLayer no longer needed? If not it should just be deleted.

File: src/main/java/li/cil/tis3d/client/gui/CodeBookGui.java [hunk 2]
```diff
@@ -652,7 +654,7 @@ private void drawProgram(final MatrixStack matrices, final int mouseX, final int
                 final List<String> lines = Arrays.asList(Constants.PATTERN_LINES.split(translatedException));
                 tooltip.addAll(lines.stream().map(Text::of).collect(Collectors.toList()));
                 renderTooltip(matrices, tooltip, mouseX, mouseY);
-                GlStateManager.disableLighting();
+                //RenderSystem._setShaderLights(null,null);
```
Comment (Sturmlilie) [inline src/main/java/li/cil/tis3d/client/gui/CodeBookGui.java:657]: Is this no longer needed, or has a workaround not been found yet?

File: src/main/java/li/cil/tis3d/client/gui/ManualGui.java [hunk 3]
```diff
@@ -103,40 +100,41 @@ public boolean mouseClicked(final double x, final double y, final int button) {
...
         scrollButton.hoverOverride = isDragging;
 
         for (int i = 0; i < ManualAPIImpl.getTabs().size() && i < MAX_TABS_PER_SIDE; i++) {
-            final ManualAPIImpl.Tab tab = ManualAPIImpl.getTabs().get(i);
-            final ImageButton button = (ImageButton)buttons.get(i);
-            GlStateManager.pushMatrix();
-            GlStateManager.translatef(button.x + 30, (float)(button.y + 4 - TAB_OVERLAP / 2), getZOffset());
-            tab.renderer.render();
-            GlStateManager.popMatrix();
+           final ManualAPIImpl.Tab tab = ManualAPIImpl.getTabs().get(i);
```
Comment (Sturmlilie) [inline src/main/java/li/cil/tis3d/client/gui/ManualGui.java:123]: indentation feels off

File: src/main/java/li/cil/tis3d/client/gui/ManualGui.java [hunk 4]
```diff
@@ -160,10 +158,12 @@ public boolean mouseScrolled(final double mx, final double my, final double valu
 
     @Override
     public boolean keyPressed(final int code, final int scancode, final int mods) {
+        if(client == null) return super.keyPressed(code, scancode, mods);
```
Comment (Sturmlilie) [inline src/main/java/li/cil/tis3d/client/gui/ManualGui.java:161]: I don't think this codestyle is used anywhere else.

File: src/main/java/li/cil/tis3d/client/gui/ManualGui.java [hunk 5]
```diff
@@ -173,10 +173,11 @@ public boolean keyPressed(final int code, final int scancode, final int mods) {
 
     @Override
     public boolean mouseClicked(final double mouseXd, final double mouseYd, final int button) {
+
+        System.out.println("Hello World!");
```
Comment (Sturmlilie) [inline src/main/java/li/cil/tis3d/client/gui/ManualGui.java:177]: leftover?

File: src/main/java/li/cil/tis3d/client/gui/ManualGui.java [hunk 6]
```diff
@@ -173,10 +173,11 @@ public boolean keyPressed(final int code, final int scancode, final int mods) {
...
         if (super.mouseClicked(mouseXd, mouseYd, button)) {
             return true;
         }
-
```
Comment (Sturmlilie) [inline src/main/java/li/cil/tis3d/client/gui/ManualGui.java:179]: spurious whitespace changes here and in other places

File: src/main/java/li/cil/tis3d/client/manual/Document.java [hunk 7]
```diff
@@ -120,37 +126,42 @@ public static int lineHeight(final TextRenderer renderer) {
...
-        GL11.glEnd();
-        GlStateManager.popMatrix();
-        GlStateManager.colorMask(true, true, true, true);
+        matrices.push();
+        matrices.translate(0, 0, 500);
+
+        final Tessellator t = Tessellator.getInstance();
+        final BufferBuilder b = t.getBuffer();
+        b.begin(VertexFormat.DrawMode.QUADS, VertexFormats.POSITION);
+        b.vertex(0, y,400);
```
Comment (Sturmlilie) [inline src/main/java/li/cil/tis3d/client/manual/Document.java:153]: spacing

File: src/main/java/li/cil/tis3d/client/mixin/WorldUnloadMinecraftClientMixin.java [hunk 8]
```diff
@@ -31,13 +31,12 @@ private void tis3d_tryDisposeCasingBlockEntities() {
...
         }
-
-        for (final BlockEntity blockEntity : world.blockEntities) {
+        /*for (final BlockEntity blockEntity : world.removeBlockEntity();) {
             if (blockEntity instanceof CasingBlockEntity) {
                 final CasingBlockEntity casingBlockEntity = (CasingBlockEntity)blockEntity;
                 final CasingImpl casing = (CasingImpl)casingBlockEntity.getCasing();
                 casing.onDisposed();
             }
-        }
```
Comment (Sturmlilie) [inline src/main/java/li/cil/tis3d/client/mixin/WorldUnloadMinecraftClientMixin.java:41]: I'm assuming there's something in the commit message about this

File: src/main/resources/fabric.mod.json [hunk 9]
```diff
@@ -4,7 +4,6 @@
     "name": "TIS-3D",
     "description": "TIS-100 inspired low-tech computing for Minecraft.",
     "version": "${version}",
-
```
Comment (Sturmlilie) [inline src/main/resources/fabric.mod.json:7]: spurious whitespace change

**Review by Sturmlilie — APPROVED** [review]
Just skimmed but lgtm; need to investigate the RenderLayerAccess thing. Sorry if I missed something that was explained in commit messages.

**Comment by fnuecke** [comment]
Thanks for the review! Didn't take the time yet, because it was still marked as draft.

FTR, I intend to try and merge the Forge branch into the Fabric one on 1.17 to get the two more in sync again; should be less painful now that Forge also uses the Mojang mappings for class names. Just not sure when I'll make the time for that :P

**Comment by Sturmlilie** [comment]
Ah my bad, I must have overlooked that it's still a draft.

**Comment by KyGost** [comment]
Is there any way I (one who doesn't have modding experience) can help finish this PR?

**Comment by lucsoft** [comment]
it would be better to migrate this first https://github.com/MightyPirates/MarkdownManual
and then this pr can go forword 

(Pull requests are open for my fork)

**Comment by Jummit** [comment]
> it would be better to migrate this first [MightyPirates/MarkdownManual](https://github.com/MightyPirates/MarkdownManual)
> and then this pr can go forword
> 
> (Pull requests are open for my fork)

I actually ported the documentation to Patchouli once, but it turns out that the book format doesn't really work well. (Text is very large and there are no code blocks)

**Comment by KyGost** [comment]
So... Looking at the current state of this PR, looks like what remains is just moving to a different library for the manual?
Using the manual in the current build shows an opening page and GUI but clicking for other pages has no effect. Further, the console is spammed with errors.

I'll have a go at migrating over to the API linked (assuming I have some time) and submit a PR to the fork.

**Comment by fnuecke** [comment]
I did start porting the extracted manual lib to fabric, not quite done yet. I might get to wrap that up next week.

Ideally this branch could be brought back closer to the forge branch in general (there were some changes to the module rendering API I'm pretty happy with), but that could come later.

**Comment by KyGost** [comment]
Ah, is there any way I can help to speed that up?

**Comment by KyGost** [comment]
Found a bug:
Right clicking ROM module shows:



Nothing special logged in console

**Comment by lucsoft** [comment]
Oh yeah the atlas is wrong probably a quick fix if you look at my git history 

(This was very Common while porting)

**Comment by fnuecke** [comment]
> Ah, is there any way I can help to speed that up?

Thanks, much appreciated, but I'm afraid not. I'm currently on vacation and didn't push before leaving 👀

**Comment by KyGost** [comment]
@fnuecke sorry to bother you, I'm also interested in a fabric port of MarkdownManual for another mod, was just curious to hear how long you're on vacation for?
(So I know whether to look for another solution or wait it out)

**Comment by KyGost** [comment]
@lucsoft 
> Oh yeah the atlas is wrong probably a quick fix if you look at my git history
> 
> (This was very Common while porting)

I've got absolutely no idea where to start looking. Any chance you can give me more of a pointer?
Trying to figure out what's going wrong and what needs to change.
My minimal knowledge leads me to suspect it is failing to grab the texture.

This is really throwing me off and seems like a gross API:
```java
client.getTextureManager().bindTexture(Textures.LOCATION_GUI_MEMORY);
drawTexture(matrices, guiX, guiY, 0, 0, GUI_WIDTH, GUI_HEIGHT);
```

We aren't saving the texture we've bound to then draw it, we are just binding into the abyss and trusting that it is remembered and presumably then trying to draw it??

**Comment by KyGost** [comment]
Figured it out, still think it is yuck. Needed this stuff:
```java
...
RenderSystem.setShaderTexture(0, Textures.LOCATION_GUI_MEMORY);
...
```

**Comment by fnuecke** [comment]
Well, this is awkward; I'm so so terribly sorry, but I completely forgot this PR existed.  I migrated the project to Architectury over the last few days, and just saw this now :/  Again, sorry not following up on this one, but I'm afraid I'd rather stick to the current MC versions now, so I'll close this one.

**Comment by lucsoft** [comment]
You moved to Architectury? NICE!

**Pull request closed without merging**